src/topology/homotopy/basic.lean

```diff
@@ -0,0 +1,394 @@
+This is defined using `homotopy_with.refl`, but with the proof filled in.
+-/
+@[simps]
+def refl (f : C(X, Y)) : homotopy f f := homotopy_with.refl f trivial
+
+/--
+Given a `homotopy f₀ f₁`, we can define a `homotopy f₁ f₀` by reversing the homotopy.
+-/
+@[simps]
+def symm {f₀ f₁ : C(X, Y)} (F : homotopy f₀ f₁) : homotopy f₁ f₀ := homotopy_with.symm F
+
+@[simp]
+lemma symm_symm {f₀ f₁ : C(X, Y)} (F : homotopy f₀ f₁) : F.symm.symm = F :=
+homotopy_with.symm_symm F
+
+/--
+Given `homotopy f₀ f₁` and `homotopy f₁ f₂`, we can define a `homotopy f₀ f₂`
+by putting the first homotopy on `[0, 1/2]` and the second on `[1/2, 1]`.
+-/
+def trans {f₀ f₁ f₂ : C(X, Y)} (F : homotopy f₀ f₁) (G : homotopy f₁ f₂) : homotopy f₀ f₂ :=
+homotopy_with.trans F G
+
+lemma trans_apply {f₀ f₁ f₂ : C(X, Y)} (F : homotopy f₀ f₁) (G : homotopy f₁ f₂)
+  (x : I × X) : (F.trans G) x =
+  if h : (x.1 : ℝ) ≤ 1/2 then
+    F (⟨2 * x.1, (unit_interval.mul_pos_mem_iff zero_lt_two).2 ⟨x.1.2.1, h⟩⟩, x.2)
+  else
+    G (⟨2 * x.1 - 1, unit_interval.two_mul_sub_one_mem_iff.2 ⟨(not_le.1 h).le, x.1.2.2⟩⟩, x.2) :=
+homotopy_with.trans_apply F G x
+
+lemma symm_trans {f₀ f₁ f₂ : C(X, Y)} (F : homotopy f₀ f₁) (G : homotopy f₁ f₂) :
+  (F.trans G).symm = G.symm.trans F.symm :=
+homotopy_with.symm_trans F G
+
+end homotopy
+
+/--
+A `homotopy_rel f₀ f₁ S` is a homotopy bewteen `f₀` and `f₁` which is fixed on the points in `S`.
+-/
+abbreviation homotopy_rel (f₀ f₁ : C(X, Y)) (S : set X) :=
+homotopy_with f₀ f₁ (λ f, ∀ x ∈ S, f x = f₀ x ∧ f x = f₁ x)
+
+namespace homotopy_rel
+
+section
+
+variables {f₀ f₁ : C(X, Y)} {S : set X}
+
+lemma eq_fst (F : homotopy_rel f₀ f₁ S) (t : I) {x : X} (hx : x ∈ S) :
+  F (t, x) = f₀ x := (F.prop t x hx).1
+
+lemma eq_snd (F : homotopy_rel f₀ f₁ S) (t : I) {x : X} (hx : x ∈ S) :
+  F (t, x) = f₁ x := (F.prop t x hx).2
+
+lemma fst_eq_snd (F : homotopy_rel f₀ f₁ S) {x : X} (hx : x ∈ S) :
+  f₀ x = f₁ x := F.eq_fst 0 hx ▸ F.eq_snd 0 hx
+
+end
+
+variables {f₀ f₁ f₂ : C(X, Y)} {S : set X}
+
+/--
+Given a map `f : C(X, Y)` and a set `S`, we can define a `homotopy_rel f f S` by setting
+`F (t, x) = f x` for all `t`. This is defined using `homotopy_with.refl`, but with the proof
+filled in.
+-/
+@[simps]
+def refl (f : C(X, Y)) (S : set X) : homotopy_rel f f S :=
+homotopy_with.refl f (λ x hx, ⟨rfl, rfl⟩)
+
+/--
+Given a `homotopy_rel f₀ f₁ S`, we can define a `homotopy_rel f₁ f₀ S` by reversing the homotopy.
+-/
+@[simps]
+def symm (F : homotopy_rel f₀ f₁ S) : homotopy_rel f₁ f₀ S :=
+{ prop' := λ t x hx, by simp [F.eq_snd _ hx, F.fst_eq_snd hx],
+  ..homotopy_with.symm F }
+
+@[simp]
+lemma symm_symm (F : homotopy_rel f₀ f₁ S) : F.symm.symm = F :=
+homotopy_with.symm_symm F
+
+/--
+Given `homotopy_rel f₀ f₁ S` and `homotopy_rel f₁ f₂ S`, we can define a `homotopy_rel f₀ f₂ S`
+by putting the first homotopy on `[0, 1/2]` and the second on `[1/2, 1]`.
+-/
+def trans (F : homotopy_rel f₀ f₁ S) (G : homotopy_rel f₁ f₂ S) : homotopy_rel f₀ f₂ S :=
```

Comment:
Right now this doesn't work with `homotopy_with.trans`, as the `P`s in `F` and `G` are different. Not sure if there is a way to make it more general so it would work?

Reply:
Hmm, I see, that is a bit awkward.

I guess the "right" way to do this is to also have "functions rel". i.e. we would write something like `homotopy_rel f0 f1 S h w0 w1`, where `h : S -> Y`, and `w0` and `w1` are proofs that `f0` and `f1` restrict to give `h`. Then it's the same `P` in all three `homotopy_rel`s.

Does that make sense? Rather than having the condition depend on `f0` and `f1`, we separately depend on their common restriction to `S`.

I'd have to think a bit if there's a way to get from here to there.

Reply:
You could try expanding the definition of `homotopy_rel` to include the existence of such an `h`, but that doesn't quite do it, as the "two different `h`s" in `F` and `G` would only be propositionally equal, not definitionally, so you'd still be in trouble trying to apply `homotopy_with.trans`.

Reply:
I might have a look at what Isabelle do for path homotopy to see how they handle the issue of there being a different predicate.

Reply:
Actually now that I think about it, if we had `homotopy_with extends homotopy`, then I think `homotopy.trans` would work here?

Reply:
Sorry, I'm not sure what you mean. What would `homotopy_rel.trans` look like?

Reply:
If we had `homotopy_with extends homotopy`, then it'd be something like

```lean
{ prop' := _,
  ..F.to_homotopy.trans G.to_homotopy }
```
I think?

Reply:
Sounds promising!

Reply:
I've pushed the changes. The doc strings are wrong/missing at the moment, but I'll add them after lunch.